### PR TITLE
KNOX-2543 - Add ability to retry failed requests

### PR DESCRIPTION
**Pull request by moresandeep**
## What changes were proposed in this pull request?
Add option to replay requests when the server unexpectedly closes the connection (not in case of socket timeouts where the server is unable to reach)
Proposed changes are adding optional service params 

1. `retryCount` - how many times should a request be retried
2. `retryNonSafeRequest` - Should an unsafe request be retries, unsafe = POST, PUT, DELETE (!GET)

example
`
    <service>
        <role>WHOAMI</role>
        <url>http://localhost:50071</url>
	<param name="retryCount" value="5"/>
    </service>
`

### CODE REVIEW

File: gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java [hunk 1]
```diff
@@ -119,6 +126,19 @@ public HttpClient createHttpClient(FilterConfig filterConfig) {
...
+    if (filterConfig.getInitParameter(PARAMETER_RETRY_COUNT) != null
+        && StringUtils
+        .isNumeric(filterConfig.getInitParameter(PARAMETER_RETRY_COUNT))) {
+      /* do we want to retry non-idempotent requests? default no */
+      boolean retryNonIdempotent = DEFAULT_PARAMETER_RETRY_NON_SAFE_REQUEST;
+      if (filterConfig.getInitParameter(PARAMETER_RETRY_NON_SAFE_REQUEST)
+          != null) {
+        retryNonIdempotent = Boolean.parseBoolean(
+            filterConfig.getInitParameter(PARAMETER_RETRY_NON_SAFE_REQUEST));
+      } builder.setRetryHandler(new DefaultHttpRequestRetryHandler(Integer
```
Comment (pzampino) [inline gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java:138]: nit: Should this be on a new line?

**Review by pzampino — APPROVED** [review]
LGTM

File: gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java [hunk 2]
```diff
@@ -119,6 +126,19 @@ public HttpClient createHttpClient(FilterConfig filterConfig) {
     // See KNOX-1530 for details
     builder.disableContentCompression();
 
+    if (filterConfig.getInitParameter(PARAMETER_RETRY_COUNT) != null
+        && StringUtils
+        .isNumeric(filterConfig.getInitParameter(PARAMETER_RETRY_COUNT))) {
```
Comment (smolnar82) [inline gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java:131]: nit: this logic worths a private boolean method IMO

File: gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java (hunk 1 above)
Comment (smolnar82) [reply gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java:138]: +1

File: gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java [hunk 3]
```diff
@@ -119,6 +126,19 @@ public HttpClient createHttpClient(FilterConfig filterConfig) {
...
+      /* do we want to retry non-idempotent requests? default no */
+      boolean retryNonIdempotent = DEFAULT_PARAMETER_RETRY_NON_SAFE_REQUEST;
+      if (filterConfig.getInitParameter(PARAMETER_RETRY_NON_SAFE_REQUEST)
+          != null) {
+        retryNonIdempotent = Boolean.parseBoolean(
+            filterConfig.getInitParameter(PARAMETER_RETRY_NON_SAFE_REQUEST));
+      } builder.setRetryHandler(new DefaultHttpRequestRetryHandler(Integer
+          .parseInt(filterConfig.getInitParameter(PARAMETER_RETRY_COUNT)),
```
Comment (smolnar82) [inline gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java:139]: `filterConfig.getInitParameter(PARAMETER_RETRY_COUNT)` could be extracted to a local variable.

**Review by smolnar82 — COMMENTED** [review]
LGTM
Minor changes are welcome.

File: gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java (hunk 1 above)
Comment (moresandeep) [reply gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java:138]: Ah right, I'll change that.

File: gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java (hunk 2 above)
Comment (moresandeep) [reply gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java:131]: yeah, I can brach this out to a private method.

File: gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java (hunk 3 above)
Comment (moresandeep) [reply gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultHttpClientFactory.java:139]: i mean we could but than I don't see a reason to create an extra variable, I'll extract it in a local variable to make it easy on the eyes :)

**Pull request merged**